### PR TITLE
New_url_formatter

### PR DESCRIPTION
**Pull request by Ni-Knight**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/XSUP-14372
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-4184
fixes: https://jira-hq.paloaltonetworks.local/browse/XSUP-15513

## Description
Whole new code for our URL formatter, checks every part of the URL and raises an error if the URL is invalid.

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [x] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

### CODE REVIEW

**Comment by Ni-Knight** [comment]
@yaakovi - Ready for CR

**Review by ChanochShayner — COMMENTED** [review]
Awesome 👑

File: Packs/CommonScripts/Scripts/FormatURL/FormatURL.py [hunk 1]
```diff
@@ -1,291 +1,549 @@
...
+        if self.original_url:
+            self.remove_leading_chars()
+
+        else:
+            raise URLError("Empty string given")
+
+        if "//" in self.modified_url[:8]:
+            """
+            The URL seems to have a scheme indicated by presence of "//"
+            """
```
Comment (DeanArbel) [inline Packs/CommonScripts/Scripts/FormatURL/FormatURL.py:65]: For inner-code comments you should use a normal comment, i.e. `#`, doctstring should usually be used in functions/classes/modules.
```suggestion
            # The URL seems to have a scheme indicated by presence of "//"
```

File: Packs/CommonScripts/Scripts/FormatURL/FormatURL.py [hunk 2]
```diff
@@ -1,291 +1,549 @@
...
-        except (ipaddress.AddressValueError, IndexError):
-            return False
+        except ValueError:
+            first_slash = -1
+
+        try:
+            if "@" in self.modified_url[:first_slash]:
+                """
+                Checks if url has '@' sign in its authority part
+                """
```
Comment (DeanArbel) [inline Packs/CommonScripts/Scripts/FormatURL/FormatURL.py:78]: Same here 
```suggestion
                # Checks if url has '@' sign in its authority part
```

File: Packs/CommonScripts/Scripts/FormatURL/FormatURL.py [hunk 3]
```diff
@@ -1,291 +1,549 @@
...
+        Parses and validates the scheme part of the URL, accepts ascii and "+", "-", "." according to standard.
+        """
+
+        index = self.base
+        scheme = ''
+
+        while self.modified_url[index].isascii() or self.modified_url[index] in ("+", "-", "."):
+            """
+            Scheme State checker
+            """
```
Comment (DeanArbel) [inline Packs/CommonScripts/Scripts/FormatURL/FormatURL.py:116]: This doesn't mean much, you can provide a more in-depth explanation, or just remove this comment

File: Packs/CommonScripts/Scripts/FormatURL/FormatURL.py [hunk 4]
```diff
@@ -1,291 +1,549 @@
...
+            if char in self.sub_delims:
+                raise URLError(f"Invalid character {char} at position {index}")
+
+            elif char == ":":
+
+                self.output += char
+                index += 1
+
+                if self.modified_url[index:index + 2] != "//":
+                    "If URL has ascii chars and ':' with no '//' it is invalid"
```
Comment (DeanArbel) [inline Packs/CommonScripts/Scripts/FormatURL/FormatURL.py:128]: Is this supposed to be a comment?
```suggestion
                    # If URL has ascii chars and ':' with no '//' it is invalid
```

File: Packs/CommonScripts/Scripts/FormatURL/FormatURL.py [hunk 5]
```diff
@@ -1,291 +1,549 @@
...
+    raw_urls = argToList(demisto.args().get('input'))
+    formatted_urls: List[str] = []
+    formatted_url = ''
+
+    for url in raw_urls:
+        try:
+            formatted_url = URLFormatter(url).output
+
+        except URLError as e:
+            demisto.info(e.__str__())
```
Comment (DeanArbel) [inline Packs/CommonScripts/Scripts/FormatURL/FormatURL.py:537]: Why is the error printed in `info` level?

**Review by DeanArbel — CHANGES_REQUESTED** [review]
Hey, please see my comments below.

File: Packs/CommonScripts/Scripts/FormatURL/FormatURL.py (hunk 4 above)
Comment (Ni-Knight) [reply Packs/CommonScripts/Scripts/FormatURL/FormatURL.py:128]: yep :)

File: Packs/CommonScripts/Scripts/FormatURL/FormatURL.py (hunk 3 above)
Comment (Ni-Knight) [reply Packs/CommonScripts/Scripts/FormatURL/FormatURL.py:116]: I will add more data, it is an entry point for the "Scheme" part of the url, we have a comment of this type to each section of the url structure, I guess adding more documentation will make it more clear, will do.

File: Packs/CommonScripts/Scripts/FormatURL/FormatURL.py [hunk 6]
```diff
@@ -1,291 +1,549 @@
...
-    if url_.startswith('http:') and not url_.startswith('http:/'):
-        url_ = url_.replace('http:', 'http://')
-    if url_.startswith('https:') and not url_.startswith('https:/'):
-        url_ = url_.replace('https:', 'https://')
-    return url_
-
-
-def remove_special_chars_from_start_and_end_of_url(url_: str) -> str:
+
+class URLCheck(object):
```
[the remark this answers is not in the document]
Comment (Ni-Knight) [reply Packs/CommonScripts/Scripts/FormatURL/FormatURL.py:32]: There is a docstring:
```
    """
    This class will build and validate a URL based on "URL Living Standard" (https://url.spec.whatwg.org)
    """    
```

File: Packs/CommonScripts/Scripts/FormatURL/FormatURL.py [hunk 7]
```diff
@@ -1,291 +1,549 @@
...
+
+        self.inside_brackets = False
+        self.port = False
+        self.query = False
+        self.fragment = False
+        self.done = False
+
+        self.sub_delims = ("!", "$", "&", "'", "(", ")", "*", "+", ",", ";", "=")
+        self.brackets = ("\"", "'", "[", "]", "{", "}", "(", ")", ",")
+        self.url_code_points = ("\u0021", "\u0024", "\u0026", "\u0027", "\u0028", "\u0029", "\u002A", "\u002B",
```
[the remark this answers is not in the document]
Comment (Ni-Knight) [reply Packs/CommonScripts/Scripts/FormatURL/FormatURL.py:52]: I was afraid to get the wrong thing (like specific quotes) this is taken straight from - https://url.spec.whatwg.org

[the remark this answers is not in the document]
Comment (Ni-Knight) [reply Packs/CommonScripts/Scripts/FormatURL/FormatURL.py:52]: we can switch it to - `('!', '$', '&', "'", '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '=', '?', '@', '_', '~')`

**Review by DeanArbel — APPROVED** [review]

**Comment by Ni-Knight** [comment]
@esharf - something was broken in the pipeline, I rerun it but in case you want to take a look (failed with no error)

**Pull request merged**